src/core/metrics.cc

```diff
@@ -109,12 +112,76 @@ future<> configure(const options& opts) {
     });
 }
 
+future<bool> set_relabel_configs(std::vector<relabel_config> relabel_configs) {
+    return impl::get_local_impl()->set_relabel_configs(relabel_configs);
+}
+
+const std::vector<relabel_config>& get_relabel_configs() {
+    return impl::get_local_impl()->get_relabel_configs();
+}
+
+
+static bool update(const relabel_config& rc, impl::metric_info& info) {
+    std::stringstream s;
+    bool first = true;
+    for (auto&& l: rc.source_labels) {
+        auto val = info.id.labels().find(l);
+        if (l != "__name__" && val == info.id.labels().end()) {
+            //If not all the labels are found nothing todo
+            return false;
+        }
+        if (first) {
+            first = false;
+        } else {
+            s << rc.separator;
+        }
+        s << ((l == "__name__") ? info.id.full_name() : val->second);
+    }
+    std::smatch match;
+    // regex_search forbid temporary strings
+    std::string tmps = s.str();
+    if (!std::regex_search(tmps, match, rc.expr.regex())) {
+        return false;
+    }
+
+    switch (rc.action) {
+        case relabel_config::relabel_action::drop:
+        case relabel_config::relabel_action::keep: {
+            info.enabled = rc.action == relabel_config::relabel_action::keep;
+            return true;
+        }
+        case relabel_config::relabel_action::report_when_empty:
+        case relabel_config::relabel_action::skip_when_empty: {
+            info.should_skip_when_empty = (rc.action == relabel_config::relabel_action::skip_when_empty) ? skip_when_empty::yes : skip_when_empty::no;
+            return false;
+        }
+        case relabel_config::relabel_action::drop_label: {
+            if (info.id.labels().find(rc.target_label) != info.id.labels().end()) {
+                info.id.labels().erase(rc.target_label);
+            }
+            return true;
+        };
+        case relabel_config::relabel_action::replace: {
+            if (!rc.target_label.empty()) {
+                std::string fmt_s = match.format(rc.replacement);
+                info.id.labels()[rc.target_label] = fmt_s;
+            }
+            return true;
+        }
+        default:
+            break;
+    }
+    return true;
+}
 
 bool label_instance::operator!=(const label_instance& id2) const {
     auto& id1 = *this;
     return !(id1 == id2);
 }
-
+std::string get_unique_id() {
+    std::random_device rd;
```

Comment:
This is extremely expensive, make it static thread_local.

Reply:
sure, keep in mind that this is super rare, and hopefully will never be used, it's for the specific scenario that someone changed a metric in such a way that would have crashed the system so this would make it possible to recover.